service/history/timerQueueTaskExecutorBase.go

```diff
@@ -138,11 +138,18 @@ func (t *timerQueueTaskExecutorBase) deleteWorkflow(
 	msBuilder workflow.MutableState,
 ) error {
 
-	if err := t.deleteCurrentWorkflowExecution(task); err != nil {
-		return err
-	}
-
-	if err := t.deleteWorkflowExecution(task); err != nil {
+	if err := t.shard.DeleteWorkflowExecution(
+		&persistence.DeleteCurrentWorkflowExecutionRequest{
+			NamespaceID: task.NamespaceID,
+			WorkflowID:  task.WorkflowID,
+			RunID:       task.RunID,
+		},
+		&persistence.DeleteWorkflowExecutionRequest{
+			NamespaceID: task.NamespaceID,
+			WorkflowID:  task.WorkflowID,
+			RunID:       task.RunID,
+		},
+	); err != nil {
 		return err
 	}
 
```

Comment:
Should we move the other two calls (deleteWorkflowVisibility and deleteWorkflowHistory) into the context too? Or fewer (just deleteCurrentWorkflowExecution since that's the lwt)?

Reply:
Is there a need where we want to delete current record and execution row, but keep history / visibility? If not, we should merge all of them into one shard context call.

Reply:
Sometimes (in the archive workflow case), we keep history:
https://github.com/temporalio/temporal/blob/master/service/history/timerQueueTaskExecutorBase.go#L228-L234
We can pass in a bool of course.

Reply:
archive is very likely to be restructured. I think we should go ahead with them separated for now. This PR's main purpose is to avoid CAS contention error. We can revisit the issue to combine delete workflow with history/visibility at a later time.

Reply:
Looking again, I'd prefer to bundle these all together. The only difference in the archive path vs. the other path is whether we delete history or not, and since we need the branch token to delete history (as far as I know), I just chose to make that argument determine whether we delete history or not.

This is easy to change when we restructure archival.